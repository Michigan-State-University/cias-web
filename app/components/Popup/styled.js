import styled from 'styled-components';
import { ZIndex } from 'theme';

export const Container = styled.div`
  position: relative;
`;

const getHorizontalPosition = (horizontalPosition) => {
  switch (horizontalPosition) {
    case 'right':
      return '0';
    case 'center':
      return '50%';
    case 'left':
      return '100%';
    default:
      break;
  }
};

const getVerticalPosition = (verticalPosition) => {
  switch (verticalPosition) {
    case 'top':
      return 'bottom: calc(100% + 5px)';
    case 'center':
      return 'top: 0';
    case 'bottom':
      return 'top: calc(100% + 5px)';
    default:
      break;
  }
};

export const PopupElement = styled.div`
  position: absolute;
  background-color: rgba(0, 0, 0, 0.8);
  padding: 8px 10px;
  border-radius: 5px;
  color: white;
  font-size: 12px;
  white-space: nowrap;
  z-index: ${ZIndex.POPUP};
  ${({ verticalPosition }) => getVerticalPosition(verticalPosition)};
  right: ${({ horizontalPosition }) =>
    getHorizontalPosition(horizontalPosition)};
`;
