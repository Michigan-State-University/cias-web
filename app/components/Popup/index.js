/**
 *
 * Popup
 *
 */

import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { Container, PopupElement } from './styled';

export const Popup = ({
  children,
  popupContent,
  verticalPosition,
  horizontalPosition,
  controlled,
  visible,
}) => {
  const [popupVisible, setPopupVisible] = useState(false);
  const togglePopup = (willBeVisible) => () => {
    if (!controlled) setPopupVisible(willBeVisible);
  };
  return (
    <Container
      onMouseEnter={togglePopup(true)}
      onMouseLeave={togglePopup(false)}
    >
      {children}
      {((!controlled && popupVisible) || visible) && (
        <PopupElement
          verticalPosition={verticalPosition}
          horizontalPosition={horizontalPosition}
        >
          {popupContent}
        </PopupElement>
      )}
    </Container>
  );
};

Popup.propTypes = {
  children: PropTypes.oneOfType([PropTypes.element, PropTypes.array])
    .isRequired,
  popupContent: PropTypes.oneOfType([PropTypes.element, PropTypes.string])
    .isRequired,
  verticalPosition: PropTypes.oneOf(['top', 'center', 'bottom']),
  horizontalPosition: PropTypes.oneOf(['left', 'center', 'right']),
  controlled: PropTypes.bool,
  visible: PropTypes.bool,
};

Popup.defaultProps = {
  verticalPosition: 'top',
  horizontalPosition: 'center',
  controlled: false,
};

export default Popup;
