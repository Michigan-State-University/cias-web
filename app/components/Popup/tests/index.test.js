/**
 *
 * Tests for Popup
 *
 * @see https://github.com/react-boilerplate/react-boilerplate/tree/master/docs/testing
 *
 */

import React from 'react';
import { render } from 'react-testing-library';
import Popup from '..';

describe('<Popup />', () => {
  it('Expect to not log errors in console', () => {
    const spy = jest.spyOn(global.console, 'error');
    render(
      <Popup popupContent={<>Test content</>}>
        <div>Test</div>
      </Popup>,
    );
    expect(spy).not.toHaveBeenCalled();
  });

  it('Should render and match the snapshot', () => {
    const {
      container: { firstChild },
    } = render(
      <Popup popupContent={<>Test content</>}>
        <div>Test</div>
      </Popup>,
    );
    expect(firstChild).toMatchSnapshot();
  });
});
