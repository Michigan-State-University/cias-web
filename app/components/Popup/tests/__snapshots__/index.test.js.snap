// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Popup /> Should render and match the snapshot 1`] = `
<div
  class="styled__Container-e4ycn-0 fbBlX"
>
  <div>
    Test
  </div>
</div>
`;
